op-program/client/l2/canon.go

```diff
@@ -0,0 +1,77 @@
+package l2
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+type BlockByHashFn func(hash common.Hash) *types.Block
+
+type CanonicalBlockHeaderOracle struct {
+	head                 *types.Header
+	hashByNum            map[uint64]common.Hash
+	earliestIndexedBlock *types.Header
+	blockByHashFn        BlockByHashFn
+}
+
+func NewCanonicalBlockHeaderOracle(head *types.Header, blockByHashFn BlockByHashFn) *CanonicalBlockHeaderOracle {
+	return &CanonicalBlockHeaderOracle{
+		head: head,
+		hashByNum: map[uint64]common.Hash{
+			head.Number.Uint64(): head.Hash(),
+		},
+		earliestIndexedBlock: head,
+		blockByHashFn:        blockByHashFn,
+	}
+}
+
+func (o *CanonicalBlockHeaderOracle) CurrentHeader() *types.Header {
+	return o.head
+}
+
+// GetHeaderByNumber walks back from the current head to the requested block number
+func (o *CanonicalBlockHeaderOracle) GetHeaderByNumber(n uint64) *types.Header {
+	if o.head.Number.Uint64() < n {
+		return nil
+	}
+
+	if o.earliestIndexedBlock.Number.Uint64() <= n {
+		// guaranteed to be cached during lookup
+		return o.blockByHashFn(o.hashByNum[n]).Header()
+	}
+
+	h := o.head
```

Comment:
This is always starting the iteration from head? We can start from the earliestIndexedBlock instead.
```suggestion
	h := o.earliestIndexedBlock
```